File: rust-api/frontendmentor/src/models/solution_model.rs
use chrono::NaiveDateTime;
use serde_derive::Serialize;

#[derive(Debug, Serialize)]
pub struct Solution {
    pub solution_id: i32,

    #[serde(rename(serialize = "createdAt"))]
    pub created_at: NaiveDateTime,

    #[serde(rename(serialize = "updatedAt"))]
    pub updated_at: NaiveDateTime,

    pub title: String,

    #[serde(rename(serialize = "repoURL"))]
    pub repo_url: String,

    #[serde(rename(serialize = "liveSiteURL"))]
    pub live_site_url: String,

    #[serde(rename(serialize = "liveSiteURL"))]
    pub image: String,

    pub tags: Vec<String>,

    pub questions: String,

    pub likes: i32,

    pub bookmarks: i32,

    #[serde(rename(serialize = "isPrivate"))]
    pub is_private: bool,

    #[serde(rename(serialize = "isBookmarked"))]
    pub is_bookmarked: bool,

    #[serde(rename(serialize = "isLiked"))]
    pub is_liked: bool,

    #[serde(rename(serialize = "isCommented"))]
    pub is_commented: bool,
}

File: client/projects/apps/frontendmentor/src/app/shared/models/solution.model.ts
import { Difficulty } from './challenge.model';
import { UserMinimal } from '@lbk/models';

export interface SolutionReport {
  a11y: number;
  htmlValidator: number;
}
/**
 * - Solution
 */
export interface Solution {
  id: string;
  challengeID: string;

  title: string;
  repoURL: string;
  liveSiteURL: string;
  screenshot: string;
  tags: string[];
  questions: string;
  languages: string[];
  createdAt: string;
  difficulty: Difficulty;
  user: UserMinimal;
  isPrivate?: boolean;
  likes: number;
  isLiked?: boolean;

  bookmarks: number;
  isBookmarked?: boolean;

  comments: number;
  isCommented?: boolean;
  report: SolutionReport;
}

/**
 * - Identify Solution
 */
export const identifySolution = (_: number, { id: solution_id }: Solution) =>
  solution_id;

/**
 * - solution Minimal
 */
export type SolutionMinimal = Pick<
  Solution,
  | 'id'
  | 'title'
  | 'screenshot'
  | 'tags'
  | 'createdAt'
  | 'likes'
  | 'comments'
  | 'bookmarks'
  | 'user'
>;

/**
 *  - Identify Solution Minimal
 */
export const identifySolutionMinimal = (
  _: number,
  { id: solution_id }: SolutionMinimal
) => solution_id;
